library_management_system_downloader/copy_and_compress_database_and_extract_tsv.py

```diff
@@ -0,0 +1,102 @@
+        '.db.xz')
+
+tsv_output_file_name = 'library_coverage_xml_and_fulltext_indicators.tsv.xz'
+
+# =============================================================================
+# Create additional settings
+# =============================================================================
+
+logging.basicConfig(level=logging.INFO)
+
+# =============================================================================
+# Set up a connection to (and tables within) the SQLite database:
+# =============================================================================
+
+# Initialize the database:
+Base = automap_base()
+sql_engine = create_engine(f'sqlite:///{sqlite_database_location}')
+Base.prepare(sql_engine, reflect=True)
+
+dois_table = Base.classes.dois_table
+library_holdings_data = Base.classes.library_holdings_data
+
+# From the sqlalchemy documentation
+# (http://docs.sqlalchemy.org/en/latest/orm/session_basics.html)
+# Create a configured "session":
+Session = sessionmaker(bind=sql_engine)
+sql_session = Session()
+
+# =============================================================================
+# Create the data directory if it doesn't already exist
+# =============================================================================
+
+if not os.path.exists(data_directory):
+    os.makedirs(data_directory)
+
+# =============================================================================
+# Extract data from the databse, and save that as a TSV
+# =============================================================================
+
+dataset_join_query = sql_session.query(
+        dois_table.doi,
+        library_holdings_data.timestamp,
+        # library_holdings_data.xml_response,
+        library_holdings_data.full_text_indicator).join(library_holdings_data)
+
+joined_dataset = pd.read_sql(
+            dataset_join_query.statement,
+            dataset_join_query.session.bind)
+
+# Sort the dataset by DOI and timestamp, so that it will always be in the
+# same order:
+joined_dataset.sort_values(
+        by=["timestamp", "doi"],
+        inplace=True)
+
+tsv_output_path = os.path.join(data_directory, tsv_output_file_name)
+
+logging.info(
+        f'Creating a TSV from the database, at "{tsv_output_path}"...')
+
+joined_dataset.to_csv(
+        path_or_buf=tsv_output_path,
+        sep='\t',
+        index=False,  # Do not write row numbers
+        compression='xz')
+
+# =============================================================================
+# Save a compressed copy of the database
+# =============================================================================
+
+database_output_path = os.path.join(
+        data_directory, compressed_database_copy_name)
+
+logging.info(
+        f'Creating a copy of the database under "'
+        f'{database_output_path}"...')
+
+with open(sqlite_database_location, "rb") as database_file:
+    with lzma.open(database_output_path, "wb") as output_file:
+        output_file.writelines(database_file)
```

Comment:
Are you sure this works? I'm a little nervous the `writelines` approach may break something.

It may be preferable to use the `xz` command line program here. You could even do it with a [`subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.run) call. `xz` [can be added](https://anaconda.org/conda-forge/xz) to the environment, if we want to ensure everyone has the utility. The two things I worry about with the above command are:

1. Is it actually xz compressing the file (could it split the bytes for a windows newline and write a unix newline)?
2. Will it be memory efficient or will large chunks of the file be loaded into memory at once?

Reply:
So, it definitely does work, at least on my system -- I tested it manually several times, where I'd download some records, run this script, open up the now-compressed database copy using `sqliteman` and compare its number of records to those in the output `tsv` and the original, uncompressed database, then download some more records and repeat the process.

But, I'm not sure of its efficiency, which it seems you're primarily asking about. I'm open to using the command-line `xz` command. The current approach of this whole chunk of code does follow the canonical `lzma` example from the [library's documentation](https://docs.python.org/3/library/lzma.html#examples), though, as I understand. `open()` [does have a `buffering` argument](https://docs.python.org/3.4/library/functions.html#open), which can be used

> to indicate the size in bytes of a fixed-size chunk buffer. When no buffering argument is given, the default buffering policy works as follows:
>    - Binary files are buffered in fixed-size chunks; the size of the buffer is chosen using a heuristic trying to determine the underlying device’s “block size” and falling back on io.DEFAULT_BUFFER_SIZE. On many systems, the buffer will typically be 4096 or 8192 bytes long.

Would `write()` be a preferable choice over `writelines()` here?

I'm willing to defer to your greater expertise with Python -- given the documentation I've cited above, which approach do you prefer?

Reply:
I'd love for a pure python implementation of memory-efficient `xz` compression a file. However, I don't see that as an [example in the docs](https://docs.python.org/3/library/lzma.html#examples). The examples don't include a concurrent read-compress-write example.

However, there is an [example](https://docs.python.org/3/library/gzip.html) in the `gzip` docs:

```python
# Example of how to GZIP compress an existing file:
import gzip
import shutil
with open('/home/joe/file.txt', 'rb') as f_in:
    with gzip.open('/home/joe/file.txt.gz', 'wb') as f_out:
        shutil.copyfileobj(f_in, f_out)
```

You should be able to apply this method but use `lzma` instead. I recommend making a `xz_compress` function, in case we want to call this multiple times.

I'm not convinced the current method you have is memory efficient and will never corrupt a file.

Reply:
You can group the with statements:

```python
with open(path_in, 'rb') as f_in, lzma.open(path_out, 'wb') as f_out:
        shutil.copyfileobj(f_in, f_out)
```

Reply:
As of 4dbfd6, I've done the following:

1. Implemented the `gzip`-documentation example you provided, but using `lzma` compression.
2. Grouped the `with` statements.
3. Wrapped the database compression into a function (called `xz_compress_a_file`)
4. Manually confirmed that the above works with the example database, by running the function on it, then opening it with `sqliteman` and running `SELECT COUNT(*) FROM dois_table` to confirm that it has the correct number of rows (currently 54).

Thanks for the suggestions!

Reply:
Did you push `4dbfd6`?

Reply:
@publicus I think you forgot to push `4dbfd6`

Reply:
Thanks for the poke. Right you are -- I did `git push origin master`, forgetting I was on a different branch. It's [here](https://github.com/greenelab/library-access/pull/12/commits/4dbfd60223a40e2e794fd516b46f61795833816b) now : )